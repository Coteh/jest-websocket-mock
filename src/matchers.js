import diff from "jest-diff";
import WS from "./websocket";

const WAIT_DELAY = 1000;
const TIMEOUT = Symbol("timoeut");

expect.extend({
  async toReceiveMessage(ws, expected) {
    const isWS = ws instanceof WS;

    if (!isWS) {
      return {
        pass: this.isNot, // always fail
        message: () =>
          this.utils.matcherHint(
            this.isNot ? ".not.toReceiveMessage" : ".toReceiveMessage",
            "WS",
            "expected"
          ) +
          "\n\n" +
          `Expected the websocket object to be a valid WS mock.\n` +
          `Received: ${typeof ws}\n` +
          `  ${this.utils.printReceived(ws)}`,
      };
    }

    const messageOrTimeout = await Promise.race([
      ws.nextMessage,
      new Promise(resolve => setTimeout(() => resolve(TIMEOUT), WAIT_DELAY)),
    ]);

    if (messageOrTimeout === TIMEOUT) {
      return {
        pass: this.isNot, // always fail
        message: () =>
          this.utils.matcherHint(
            this.isNot ? ".not.toReceiveMessage" : ".toReceiveMessage",
            "WS",
            "expected"
          ) +
          "\n\n" +
          `Expected the websocket server to receive a message,\n` +
          `but it didn't receive anything in ${WAIT_DELAY}ms.`,
      };
    }
    const received = messageOrTimeout;

    const pass = this.equals(received, expected);

    const message = pass
      ? () =>
          this.utils.matcherHint(".not.toReceiveMessage", "WS", "expected") +
          "\n\n" +
          `Expected the next received message to not equal:\n` +
          `  ${this.utils.printExpected(expected)}\n` +
          `Received:\n` +
          `  ${this.utils.printReceived(received)}`
      : () => {
          const diffString = diff(expected, received, { expand: this.expand });
          return (
            this.utils.matcherHint(".toReceiveMessage", "WS", "expected") +
            "\n\n" +
            `Expected the next received message to equal:\n` +
            `  ${this.utils.printExpected(expected)}\n` +
            `Received:\n` +
            `  ${this.utils.printReceived(received)}\n\n` +
            `Difference:\n\n${diffString}`
          );
        };

    return {
      actual: received,
      expected,
      message,
      name: "toReceiveMessage",
      pass,
    };
  },
});
