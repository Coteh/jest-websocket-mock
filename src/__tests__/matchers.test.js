import WS from "../websocket";
import "../matchers";

let server, client;
beforeEach(async () => {
  server = new WS("ws://localhost:1234");
  client = new WebSocket("ws://localhost:1234");
  await server.connected;
});

afterEach(() => {
  WS.clean();
});

describe(".toReceiveMessage", () => {
  it("passes when the websocket server receives the expected message", async () => {
    client.send("hello there");
    await expect(server).toReceiveMessage("hello there");
  });

  it("passes when the websocket server receives the expected JSON message", async () => {
    const jsonServer = new WS("ws://localhost:9876", { jsonProtocol: true });
    const jsonClient = new WebSocket("ws://localhost:9876");
    await jsonServer.connected;
    jsonClient.send(`{"answer":42}`);
    await expect(jsonServer).toReceiveMessage({ answer: 42 });
  });

  it("fails when called with an expected argument that is not a valid WS", async () => {
    expect.hasAssertions();
    await expect(
      expect("boom").toReceiveMessage("hello there")
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("fails when the WS server does not receive the expected message", async () => {
    expect.hasAssertions();
    await expect(
      expect(server).toReceiveMessage("hello there")
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("fails when the WS server receives a different message", async () => {
    expect.hasAssertions();
    client.send("hello there");
    await expect(
      expect(server).toReceiveMessage("HI!")
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("fails when expecting a JSON message but the server is not configured for JSON protocols", async () => {
    expect.hasAssertions();
    client.send(`{"answer":42}`);
    await expect(
      expect(server).toReceiveMessage({ answer: 42 })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe(".not.toReceiveMessage", () => {
  it("passes when the websocket server doesn't receive the expected message", async () => {
    client.send("hello there");
    await expect(server).not.toReceiveMessage("What's up?");
  });

  it("fails when called with an expected argument that is not a valid WS", async () => {
    expect.hasAssertions();
    await expect(
      expect("boom").not.toReceiveMessage("hello there")
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("fails when the WS server doesn't receive any messages", async () => {
    expect.hasAssertions();
    await expect(
      expect(server).not.toReceiveMessage("hello there")
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("fails when the WS server receives the un-expected message", async () => {
    expect.hasAssertions();
    client.send("hello there");
    await expect(
      expect(server).not.toReceiveMessage("hello there")
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
