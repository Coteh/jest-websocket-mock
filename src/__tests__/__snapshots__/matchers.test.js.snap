// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toReceiveMessage fails when called with an expected argument that is not a valid WS 1`] = `
"[2mexpect([22m[31mWS[39m[2m).not.toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket object to be a valid WS mock.
Received: string
  [31m\\"boom\\"[39m"
`;

exports[`.not.toReceiveMessage fails when the WS server doesn't receive any messages 1`] = `
"[2mexpect([22m[31mWS[39m[2m).not.toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket server to receive a message,
but it didn't receive anything in 1000ms."
`;

exports[`.not.toReceiveMessage fails when the WS server receives the un-expected message 1`] = `
"[2mexpect([22m[31mWS[39m[2m).not.toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the next received message to not equal:
  [32m\\"hello there\\"[39m
Received:
  [31m\\"hello there\\"[39m"
`;

exports[`.toReceiveMessage fails when called with an expected argument that is not a valid WS 1`] = `
"[2mexpect([22m[31mWS[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket object to be a valid WS mock.
Received: string
  [31m\\"boom\\"[39m"
`;

exports[`.toReceiveMessage fails when expecting a JSON message but the server is not configured for JSON protocols 1`] = `
"[2mexpect([22m[31mWS[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the next received message to equal:
  [32m{\\"answer\\": 42}[39m
Received:
  [31m\\"{\\\\\\"answer\\\\\\":42}\\"[39m

Difference:

  Comparing two different types of values. Expected [32mobject[39m but received [31mstring[39m."
`;

exports[`.toReceiveMessage fails when the WS server does not receive the expected message 1`] = `
"[2mexpect([22m[31mWS[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket server to receive a message,
but it didn't receive anything in 1000ms."
`;

exports[`.toReceiveMessage fails when the WS server receives a different message 1`] = `
"[2mexpect([22m[31mWS[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the next received message to equal:
  [32m\\"HI!\\"[39m
Received:
  [31m\\"hello there\\"[39m

Difference:

[32m- Expected[39m
[31m+ Received[39m

[32m- HI![39m
[31m+ hello there[39m"
`;
